src/x/debug/debug.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package debug
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// Source provides functions for fetching debug information from a single debug source.
+type Source interface {
+	// Write writes it's debug information into the provided writer.
+	Write(w io.Writer) error
+}
+
+// ZipWriter aggregates sources and writes them in a zip file.
+type ZipWriter interface {
+	// WriteZip writes a ZIP file in the provided writer.
+	// The archive contains the dumps of all sources in separate files.
+	WriteZip(io.Writer) error
+	// RegisterSource adds a new source to the produced archive.
+	RegisterSource(string, Source) error
+	// HTTPHandler sends out the ZIP file as raw bytes.
+	HTTPHandler() http.Handler
+	// RegisterHandler wires the HTTPHandlerFunc with the given router.
+	RegisterHandler(string, *http.ServeMux) error
+}
+
+type zipWriter struct {
+	sources map[string]Source
+}
+
+// NewZipWriter returns an instance of an ZipWriter. The passed prefix
+// indicates the folder where to save the zip files.
+func NewZipWriter() ZipWriter {
```

Comment:
Could we add another constructor function: `NewZipWriterWithDefaultSources` that basically creates a zip writer, then creates/registers the CPU/heap/host sources? That way when we want to use this package and register the http endpoint it basically just becomes:

```golang
zipWriter, err := debug.NewZipWriterWithDefaultSources()
if err != nil {
   ...
}
// Register any custom sources here.
...
handler := zipWriter.HTTPHandler()
...
```

Reply:
Done!